airbyte-workers/src/main/java/io/airbyte/workers/temporal/PartialSuccessTemporalAttemptExecution.java

```diff
@@ -0,0 +1,221 @@
+                                         CheckedConsumer<Path, IOException> jobRootDirCreator,
+                                         CancellationHandler cancellationHandler,
+                                         Supplier<String> workflowIdProvider,
+                                         Predicate<OUTPUT> shouldAttemptAgainPredicate,
+                                         BiFunction<INPUT, OUTPUT, INPUT> computeNextAttemptInputFunction,
+                                         int maxRetriesCount) {
+    this(
+        workspaceRoot,
+        jobRunConfig,
+        workerSupplier,
+        initialInputSupplier,
+        cancellationHandler,
+        shouldAttemptAgainPredicate,
+        computeNextAttemptInputFunction,
+        maxRetriesCount,
+        (Path workspaceRootArg,
+         JobRunConfig jobRunConfigArg,
+         CheckedSupplier<Worker<INPUT, OUTPUT>, Exception> workerSupplierArg,
+         Supplier<INPUT> initialInputSupplierArg,
+         // for testing.
+         CancellationHandler e) -> new TestTemporalAttemptExecutionFactory<INPUT, OUTPUT>(mdcSetter, jobRootDirCreator, workflowIdProvider)
+             .create(workspaceRootArg, jobRunConfigArg, workerSupplierArg, initialInputSupplierArg, e));
+  }
+
+  @VisibleForTesting
+  PartialSuccessTemporalAttemptExecution(Path workspaceRoot,
+                                         JobRunConfig jobRunConfig,
+                                         CheckedSupplier<Worker<INPUT, OUTPUT>, Exception> workerSupplier,
+                                         Supplier<INPUT> initialInputSupplier,
+                                         CancellationHandler cancellationHandler,
+                                         Predicate<OUTPUT> shouldAttemptAgainPredicate,
+                                         BiFunction<INPUT, OUTPUT, INPUT> computeNextAttemptInputFunction,
+                                         int maxRetriesCount,
+                                         TemporalAttemptExecutionFactory<INPUT, OUTPUT> temporalAttemptExecutionFactory) {
+    this.workspaceRoot = workspaceRoot;
+    this.jobRunConfig = jobRunConfig;
+    this.workerSupplier = workerSupplier;
+    this.inputSupplier = initialInputSupplier;
+    this.cancellationHandler = cancellationHandler;
+    this.shouldAttemptAgainPredicate = shouldAttemptAgainPredicate;
+    this.computeNextAttemptInputFunction = computeNextAttemptInputFunction;
+    this.maxRetriesCount = maxRetriesCount;
+    this.temporalAttemptExecutionFactory = temporalAttemptExecutionFactory;
+  }
+
+  @Override
+  public List<OUTPUT> get() {
+    INPUT input = inputSupplier.get();
+    final AtomicReference<OUTPUT> lastOutput = new AtomicReference<>();
+    List<OUTPUT> outputCollector = new ArrayList<>();
+
+    int i = 0;
+    while (true) {
+      if (i >= maxRetriesCount) {
+        LOGGER.info("Max retries reached: {}", i);
+        break;
+      }
+
+      final boolean hasLastOutput = lastOutput.get() != null;
```

Comment:
what if a source outputs no data? this would falsely retry right?

Reply:
`OUTPUT` in this case is the `StandardSyncSummary`. So if the the worker fails to output a `StandardSyncSummary` it is definitely a failure case.